### PR TITLE
add support to set UiAutomator Congfigurator values 

### PR DESCRIPTION
**Pull request by truebit**
This is a sub PR for appium/appium#6561

### CODE REVIEW

File: bootstrap/pom.xml [hunk 1]
```diff
@@ -31,8 +31,9 @@
             <plugin>
                 <artifactId>maven-compiler-plugin</artifactId>
                 <configuration>
-                    <source>1.6</source>
-                    <target>1.6</target>
+                    <source>1.7</source>
```
Comment (jlipps) [inline bootstrap/pom.xml]: what are the ramifications of this upgrade?

File: bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            if (params.containsKey("value")) {
+                value = (Integer) params.get("value");
+            }
+            String methodName = ((String) params.get("method"));
+            //TODO: use reflection to invoke method would be more expandable; but Configurator is singleton
+            Configurator configurator = Configurator.getInstance();
+            switch (methodName) {
+                case "setActionAcknowledgmentTimeout":
+                    if (value == 0) // set to default when no value or zero
+                        value = 3000;
```
Comment (jlipps) [inline bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java]: should we make these defaults constants up top so they're easier to see and change?

File: bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    public AndroidCommandResult execute(AndroidCommand command) throws JSONException {
+        if (API_18) {
+            final Hashtable<String, Object> params = command.params();
+            int value = 0;
+            if (params.containsKey("value")) {
+                value = (Integer) params.get("value");
+            }
+            String methodName = ((String) params.get("method"));
+            //TODO: use reflection to invoke method would be more expandable; but Configurator is singleton
+            Configurator configurator = Configurator.getInstance();
```
Comment (jlipps) [inline bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java]: Configurator is cool, I wasn't familiar with it. Are we sure that it is available in the older versions of UiAutomator?

File: bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                    if (value == 0) // set to default when no value
+                        value = 10000;
+                    configurator.setWaitForSelectorTimeout(value);
+                    break;
+                default:
+                    return getErrorResult("'configurator' command must contain 'type' key");
+            }
+            return getSuccessResult(value);
+        } else {
+            return getErrorResult("Device does not support API >= 18!");
```
Comment (jlipps) [inline bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java]: oh, guess you're saying that it's API 18 and up for Configurator support?

**Comment by jlipps** [comment]
couple comments then LGTM. would also love to see a test though!

File: bootstrap/pom.xml (hunk 1 above)
Comment (truebit) [inline bootstrap/pom.xml]: I used String switch statement in code, it needs Java 1.7

File: bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java (hunk 3 above)
Comment (truebit) [inline bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java]: Configurator class was first introduced in API 18 in api document.

File: bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java (hunk 4 above)
Comment (truebit) [inline bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java]: Yep, I copied this line from somewhere in this repo. did not check for the message content 😅

Comment (truebit) [inline bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java]: Yep, I copied this line from somewhere in this repo. did not check for the message content 😅

Comment (imurchie) [inline bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java]: I would prefer the error condition to be at the top, and the rest of the method be outside of the conditional. That is:

```java
if (!API_18) {
  return getErrorResult("Device does not support API >= 18!");
}

final Hashtable<String, Object> params = command.params();
// ...
```

File: bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                    if (value == 0) // set to default when no value or zero
+                        value = 3000;
+                    configurator.setActionAcknowledgmentTimeout(value);
+                    break;
+                case "setKeyInjectionDelay":
+                    //KeyInjectionDelay default is 0
+                    configurator.setKeyInjectionDelay(value);
+                    break;
+                case "setScrollAcknowledgmentTimeout":
+                    if (value == 0) // set to default when no value
```
Comment (imurchie) [inline bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java]: Even though there is only one line, our style is to prefer blocks to have brackets.

Comment (truebit) [inline bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java]: done in https://github.com/appium/appium-android-bootstrap/pull/25/commits/4a35db8c76afa5fcaffb92a5b9adbc0b4c042f49

File: bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java (hunk 4 above)
Comment (truebit) [inline bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java]: done in https://github.com/appium/appium-android-bootstrap/pull/25/commits/e372568409227c95038150ac352617fc2251c28a

File: bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java (hunk 2 above)
Comment (truebit) [inline bootstrap/src/io/appium/android/bootstrap/handler/ConfiguratorHandler.java]: done in https://github.com/appium/appium-android-bootstrap/pull/25/commits/d6dbb312f9ca38465ee1823c576e17545bf04945

**Comment by truebit** [comment]
changed set-to-default logic, now using negative value instead of 0 in https://github.com/appium/appium-android-bootstrap/pull/25/commits/9a060501278c103c5bc286178d793b8eebe39771

**Comment by imurchie** [comment]
Does this run locally? Not sure why it is not being built in Travis.

**Comment by truebit** [comment]
> Does this run locally? Not sure why it is not being built in Travis.

According to the TravisCI log, this project requires a very old node version 0.12. Now the major version of node is 6 now. So maybe nvm cannot find that old version of node.

**Comment by truebit** [comment]
> Does this run locally? Not sure why it is not being built in Travis.

Ah, I know. I thought the Android project was working on maven via README. So I just changed java version in pom; Although I could only build success locally using ant with `ant build -Djava.source=7 -Djava.target=7`

I committed https://github.com/appium/appium-android-bootstrap/pull/25/commits/d24c07e71e600778658e73038f72d31555bea853 to fix this build error

**Comment by truebit** [comment]
Now the `ant-build` task in gulp passed, but now executing `npm install` still fail due to network problem in Travis:

> npm ERR! argv "/home/travis/.nvm/versions/node/v0.12.15/bin/node" "/home/travis/.nvm/versions/node/v0.12.15/bin/npm" "install"
> npm ERR! node v0.12.15
> npm ERR! npm  v2.15.1

**Comment by imurchie** [comment]
Yes, Travis was struggling yesterday.

All seems well, now. Can you squash into a single commit, and we can proceed? Thanks!

**Comment by truebit** [comment]
squash done

**Comment by imurchie** [comment]
Thanks for all your work, and patience in getting this in!

**Pull request merged**